go/libraries/doltcore/sqle/dolt_log_table_function.go

```diff
@@ -267,104 +266,106 @@ func (ltf *LogTableFunction) WithExpressions(expression ...sql.Expression) (sql.
-		}
+	notRevisionStrs, err := mustExpressionsToString(ltf.ctx, ltf.notRevisionExprs)
+	if err != nil {
+		return err
 	}
 
-	if ltf.revisionExpr != nil && ltf.secondRevisionExpr != nil {
-		if strings.HasPrefix(revisionStr, "^") && strings.HasPrefix(secondRevisionStr, "^") {
-			return ltf.invalidArgDetailsErr(ltf.revisionExpr, "both revisions cannot contain '^'")
+	for i, revisionStr := range revisionStrs {
+		if !types.IsText(ltf.revisionExprs[i].Type()) {
+			return ltf.invalidArgDetailsErr(ltf.revisionExprs[i].String())
 		}
-		if !strings.HasPrefix(revisionStr, "^") && !strings.HasPrefix(secondRevisionStr, "^") {
-			return ltf.invalidArgDetailsErr(ltf.revisionExpr, "one revision must contain '^' if two revisions provided")
+		if strings.Contains(revisionStr, "..") && (len(revisionStrs) > 1 || ltf.notRevisionExprs != nil || ltf.notRevisionStrs != nil) {
+			return ltf.invalidArgDetailsErr("revision cannot contain '..' or '...' if multiple revisions exist")
 		}
 	}
 
-	if len(ltf.notRevision) > 0 {
-		if ltf.revisionExpr == nil && ltf.secondRevisionExpr == nil {
-			return ltf.invalidArgDetailsErr(ltf.revisionExpr, "must have revision in order to use --not")
-		}
-		if ltf.revisionExpr != nil && (strings.Contains(revisionStr, "..") || strings.HasPrefix(revisionStr, "^")) {
-			return ltf.invalidArgDetailsErr(ltf.revisionExpr, "cannot use --not if dots or '^' present in revision")
+	for i, notRevisionStr := range notRevisionStrs {
+		if !types.IsText(ltf.notRevisionExprs[i].Type()) {
+			return ltf.invalidArgDetailsErr(ltf.notRevisionExprs[i].String())
 		}
-		if ltf.secondRevisionExpr != nil && strings.HasPrefix(secondRevisionStr, "^") {
-			return ltf.invalidArgDetailsErr(ltf.secondRevisionExpr, "cannot use --not if '^' present in second revision")
+		if strings.Contains(notRevisionStr, "..") {
+			return ltf.invalidArgDetailsErr("revision cannot contain both '..' or '...' and '^'")
 		}
-		if strings.Contains(ltf.notRevision, "..") {
-			return sql.ErrInvalidArgumentDetails.New(ltf.Name(), fmt.Sprintf("%s - %s", ltf.notRevision, "--not revision cannot contain '..'"))
+	}
+	for _, notRevStr := range ltf.notRevisionStrs {
+		if strings.Contains(notRevStr, "..") {
+			return ltf.invalidArgDetailsErr("--not revision cannot contain '..'")
 		}
-		if strings.HasPrefix(ltf.notRevision, "^") {
-			return sql.ErrInvalidArgumentDetails.New(ltf.Name(), fmt.Sprintf("%s - %s", ltf.notRevision, "--not revision cannot contain '^'"))
+		if strings.Contains(notRevStr, "^") {
+			return ltf.invalidArgDetailsErr("--not revision cannot contain '^'")
 		}
 	}
 
 	return nil
 }
 
+func mustExpressionsToString(ctx *sql.Context, expr []sql.Expression) ([]string, error) {
+	var valStrs []string
+
+	for _, ex := range expr {
+		valStr, err := expressionToString(ctx, ex)
+		if err != nil {
+			return nil, err
+		}
+
+		valStrs = append(valStrs, valStr)
+	}
+
+	return valStrs, nil
+}
+
+func expressionToString(ctx *sql.Context, expr sql.Expression) (string, error) {
```

Comment:
Do we have any tests which test actually expressions? All of our tests inputs are strings, so there is no evaluation actually being done. Please add a test which has a select or a function call in it as an argument to dolt_log()

Reply:
nm. discussed in person that @CommitA is evaled.